src/backend/access/appendonly/appendonlyam_handler.c

```diff
@@ -1050,12 +1050,184 @@ appendonly_relation_copy_for_cluster(Relation OldHeap, Relation NewHeap,
+{
+	AppendOnlyExecutorReadBlock	*readblock;
+	FileSegInfo	*seginfo;
+	int64	rows_prev_segments; /* row count from prevoius segments (except current segment) */
+
+	readblock = &scan->executorReadBlock;
+
+	/* Need to open a new segment file. */
+	if (scan->aos_need_new_segfile)
+	{
+		if (!SetNextFileSegForRead(scan))
+			return false;
+		readblock->segmentRowsScanned = 0;
+	}
+
+	/* No more segments left. */
+	if (scan->aos_done_all_segfiles)
+	{
+		elog(DEBUG2, "No more segments left.");
+		return false;
+	}
+
+	if (readblock->totalRowsScanned > targetrow)
+	{
+		elog(ERROR, "Analyze block inconsistency: total rows scanned %lu, target row %lu",
+			readblock->totalRowsScanned, targetrow);
+	}
+
+	/* Look up for the target row in a current segment. */
+	seginfo = scan->aos_segfile_arr[scan->aos_segfiles_processed - 1];
+	rows_prev_segments = readblock->totalRowsScanned - readblock->segmentRowsScanned;
+	if (targetrow <= rows_prev_segments + seginfo->total_tupcount)
+	{
+		return true;
+	}
+
+	/* Look up for the target row among other segments. */
+	while (true)
+	{
+		readblock->totalRowsScanned += seginfo->total_tupcount - readblock->segmentRowsScanned;
+
+		scan->aos_need_new_segfile = true;
+		CloseScannedFileSeg(scan);
+		elog(DEBUG2, "Skipping current segment: segno %d", readblock->segmentFileNum);
+		if (!getNextBlock(scan))
+		{
+			elog(DEBUG3, "No more varblocks left in a current segment.");
+			return false;
+		}
+
+		/* Skip a segment until it contains the target row. */
+		seginfo = scan->aos_segfile_arr[scan->aos_segfiles_processed - 1];
+		if (readblock->totalRowsScanned + seginfo->total_tupcount >= targetrow)
+		{
+			break;
+		}
+	}
+
+	return true;
+}
+
 static bool
 appendonly_scan_analyze_next_block(TableScanDesc scan, BlockNumber blockno,
 							   BufferAccessStrategy bstrategy)
 {
+	int64 targetrow;
+	int64 varblockFirstRow;	/* var block row position with respect to previous segments */
+	int64 varblockRemainingRows;
+	FileSegInfo *seginfo;
+	AppendOnlyExecutorReadBlock *readblock;
+	TupleTableSlot *slot;
+	bool found;
+	int segno;
+	int64 totalrows;
+	int64 totalRemainingRows;
+
 	AppendOnlyScanDesc aoscan = (AppendOnlyScanDesc) scan;
-	aoscan->targetTupleId = blockno;
+
+	/* Convert block number to a target row in an ordered sequence of appendonly segments. */
+	targetrow = blockno * aoscan->analyze_block_size;
+
+	/* This should never happen */
+	if (targetrow > INT64_MAX)
+	{
+		elog(ERROR, "Target row is out on range");
+	}
+
+	/* Find a segment with a target row (skip other ones without scan and increment counters) */
+	if (!appendonly_analyze_find_segment(aoscan, targetrow))
+	{
+		elog(DEBUG3, "No segment exists for the target row %lu", targetrow);
+		return false;
+	}
+
+	/* Calculate rows to scan in a current block. */
+	readblock = &aoscan->executorReadBlock;
+	totalrows = 0;
+	for (segno = 0; segno < aoscan->aos_total_segfiles; segno++)
+	{
+		totalrows += aoscan->aos_segfile_arr[segno]->total_tupcount;
```

Comment:
We should use `GetAOCSSSegFilesTotals()` under snapshot for it instead.

Reply:
fixed in https://github.com/arenadata/gpdb/pull/135/commits/eff85e01962f23eeb74c96e167ac2018c40d663a